src/encoder/encoderopus.cpp

```diff
@@ -0,0 +1,439 @@
+// encoderopus.cpp
+// Create on August 15th 2017 by Palakis
+
+#include <QByteArray>
+#include <QMapIterator>
+
+#include "encoder/encoderopussettings.h"
+#include "engine/sidechain/enginesidechain.h"
+#include "util/logger.h"
+
+#include "encoder/encoderopus.h"
+
+namespace {
+// From libjitsi's Opus encoder:
+// 1 byte TOC + maximum frame size (1275)
+// See https://tools.ietf.org/html/rfc6716#section-3.2
+const int kMaxOpusBufferSize = 1+1275;
+// Opus frame duration in milliseconds. Fixed to 60ms
+const int kOpusFrameMs = 60;
+
+const mixxx::Logger kLogger("EncoderOpus");
+}
+
+EncoderOpus::EncoderOpus(EncoderCallback* pCallback)
+    : m_bitrate(0),
+      m_bitrateMode(0),
+      m_channels(0),
+      m_samplerate(0),
+	  m_readRequired(0),
+      m_pCallback(pCallback),
+      m_pFifoBuffer(nullptr),
+      m_pFifoChunkBuffer(nullptr),
+      m_pOpus(nullptr),
+      m_header_write(false),
+      m_packetNumber(0),
+      m_granulePos(0) {
+    m_opusComments.insert("ENCODER", "mixxx/libopus");
+    m_pOpusDataBuffer = new unsigned char[kMaxOpusBufferSize]();
+    ogg_stream_init(&m_oggStream, rand());
+}
+
+EncoderOpus::~EncoderOpus() {
+    if (m_pOpus)
+        opus_encoder_destroy(m_pOpus);
+
+    ogg_stream_clear(&m_oggStream);
+    delete[] m_pOpusDataBuffer;
+    delete m_pFifoChunkBuffer;
+    delete m_pFifoBuffer;
+}
+
+void EncoderOpus::setEncoderSettings(const EncoderSettings& settings) {
+    m_bitrate = settings.getQuality();
+    m_bitrateMode = settings.getSelectedOption(EncoderOpusSettings::BITRATE_MODE_GROUP);
+    switch (settings.getChannelMode()) {
+        case EncoderSettings::ChannelMode::MONO:
+            m_channels = 1;
+            break;
+        case EncoderSettings::ChannelMode::STEREO:
+            m_channels = 2;
+            break;
+        case EncoderSettings::ChannelMode::AUTOMATIC:
+            m_channels = 2;
+            break;
+    }
+}
+
+int EncoderOpus::initEncoder(int samplerate, QString errorMessage) {
+    (void)errorMessage;
+    int result;
+
+    if (samplerate != 48000) {
+        kLogger.warning() << "initEncoder failed: samplerate not supported by Opus";
+
+        QString invalidSamplerate = QObject::tr(
```

Comment:
Almost the same message is used in ShoutConnection, even using the same "magic numbers". Please try to minimize redundant code and decide according to the dependencies were to place shared code and constants.